neuralmonkey/decoders/beam_search_decoder.py

```diff
@@ -1,442 +1,531 @@
+        # ``search_state``?
+
         return BeamSearchLoopState(
-            bs_state=self._search_state,
-            bs_output=output_ta,
-            decoder_loop_state=dec_ls)
-
-    def _decoding_loop(self) -> BeamSearchOutput:
-        # collect attention objects
-        beam_body = self.get_body()
-
-        initial_loop_state = self.get_initial_loop_state()
-
-        def cond(*args) -> tf.Tensor:
-            bsls = BeamSearchLoopState(*args)
-            return tf.less(
-                bsls.decoder_loop_state.feedables.step - 1, self._max_steps)
-
-        # First step has to be run manually because while_loop needs the same
-        # shapes between steps and the first beam state is not beam-sized, but
-        # just a single state.
-        #
-        # When running ensembles, we want to provide
-        # ensembled logprobs to the beam_body before manually running
-        # the first step
-        next_bs_loop_state = tf.cond(
-            cond(*initial_loop_state),
-            lambda: beam_body(*initial_loop_state),
-            lambda: initial_loop_state)
-
-        final_state = tf.while_loop(cond, beam_body, next_bs_loop_state)
-        dec_loop_state = final_state.decoder_loop_state
-        bs_state = final_state.bs_state
-
-        # Because the "batch" dimension is dynamic and the TensorArrays
-        # are empty during initial call during ensembling, we have
-        # to define default values (the runner itself has to handle them,
-        # if needed)
-        # NOTE this code might change in the future, if tf.cond gets replaced
-        # by something more reasonable
-        scores = tf.cond(
-            tf.equal(final_state.bs_output.scores.size(), 0),
-            lambda: tf.fill([1, self.batch_size, self.beam_size], 0.0),
-            final_state.bs_output.scores.stack)
-        parent_ids = tf.cond(
-            tf.equal(final_state.bs_output.parent_ids.size(), 0),
-            lambda: tf.fill([1, self.batch_size, self.beam_size], 0),
-            final_state.bs_output.parent_ids.stack)
-        token_ids = tf.cond(
-            tf.equal(final_state.bs_output.token_ids.size(), 0),
-            lambda: tf.fill([1, self.batch_size, self.beam_size], 0),
-            final_state.bs_output.token_ids.stack)
+            search_state=search_state,
+            search_results=search_results,
+            decoder_loop_state=dec_next_ls)
+
+    def loop_continue_criterion(self, *args) -> tf.Tensor:
+        """Decide whether to break out of the while loop.
+
+        The criterion for stopping the loop is that either all hypotheses are
+        finished or a maximum number of steps has been reached. Here the number
+        of steps is the number of steps of the underlying decoder minus one
+        because this function is evaluated after the decoder step has been
+        called and its step has been incremented.
+
+        TODO(@varisd) check this please
```

Comment:
V poradku (akorat si nejsem jisty, ze je z toho uplne jasne, ze BS dekoder je o krok pozadu za parentem a ze to je kvuli tomu, ze parent_body se vola az na konci bs_body, coz je kvuli ensemblum)

Reply:
ok